src/shared/lock-util.c

```diff
@@ -115,3 +115,32 @@ void release_lock_file(LockFile *f) {
         f->fd = safe_close(f->fd);
         f->operation = 0;
 }
+
+int unposix_lock(int fd, int operation) {
+        int cmd, type;
+
+        assert(fd >= 0);
+
+        cmd = (operation & LOCK_NB) ? F_OFD_SETLK : F_OFD_SETLKW;
+
+        switch (operation & ~LOCK_NB) {
+                case LOCK_EX:
+                        type = F_WRLCK;
+                        break;
+                case LOCK_SH:
+                        type = F_RDLCK;
+                        break;
+                case LOCK_UN:
+                        type = F_UNLCK;
+                        break;
+                default:
+                        assert_not_reached();
+        }
+
+        return RET_NERRNO(fcntl(fd, cmd, &(struct flock) {
+                .l_type = type,
+                .l_whence = SEEK_SET,
+                .l_start = 0,
+                .l_len = 0,
+        }));
+}
```

Comment:
btw, the docs say fcntl() might return either EACCESS or EAGAIN in case a lock is already taken and suggests handling it the same.

BSD flock() only returns EAGAIN in case the lock is already taken. Hence maybe add:

```c
if (IN_SET(r, -EAGAIN, -EACCESS)) /* treat EACCESS/EAGAIN the same as per man page */
        return -EAGAIN;
return r;
```